idseq_dag/util/count.py

```diff
@@ -18,86 +17,40 @@ class ReadCountingMode(Enum):
-    Supports fastq and SINGLE-LINE fasta formats.
-    TODO: add support for m8 files here once the relevant steps are added in the pipeline engine.
+def count_reads(filename):
     '''
-    read_count = None
-    if file_format in ["fq", "fastq"]:
-        # Each read consists of exactly 4 lines, by definition.
-        assert line_count % 4 == 0, "File does not follow fastq format"
-        read_count = line_count // 4
-    elif file_format in ["fa", "fasta"]:
-        # Each read consists of exactly 2 lines (identifier and sequence).
-        # ASSUMES the format is SINGLE-LINE fasta files, where each read's sequence
-        # is on a single line rather than being spread over multiple lines.
-        # This is fine for now, because we only count fasta files we generated
-        # ourselves, following the single-line format.
-        assert line_count % 2 == 0, "File does not follow single-line fasta format"
-        read_count = line_count // 2
-    else:
-        read_count = line_count
-    return read_count
-
-
-def _reads2lines(read_count, file_format):
+    Count reads in a given FASTA or FASTQ file.
     '''
-    Convert read count to line count based on file format.
-    Currently supports fastq or SINGLE-LINE fasta.
-    '''
-    line_count = None
-    if file_format in ["fq", "fastq"]:
-        line_count = 4 * read_count
-    elif file_format in ["fa", "fasta"]:
-        # Assumes the format is single-line fasta rather than multiline fasta
-        line_count = 2 * read_count
-    return line_count
+    libmagic_cmd = ["file", "--dereference", "--brief", "--mime-type", filename]
+    is_gzipped = True if run(libmagic_cmd, stdout=PIPE).stdout.startswith(b"application/gzip") else False
+    with gzip.open(filename) if is_gzipped else open(filename, mode="rb") as fmt_fh:
+        first_char = fmt_fh.read(1).decode()[0]
+    with open(filename, "rb") as fh:
+        if first_char == ">":
+            cmd = "grep -c '^>'"
+            if is_gzipped:
+                cmd = "gunzip | " + cmd
+            return int(run(cmd, stdin=fh, stdout=PIPE, check=True, shell=True).stdout)
+        elif first_char == "@":
+            cmd = "wc -l"
+            if is_gzipped:
+                cmd = "gunzip | " + cmd
+            num_lines = int(run(cmd, stdin=fh, stdout=PIPE, check=True, shell=True).stdout)
+            if num_lines % 4 != 0:
+                raise InvalidInputFileError("File does not follow fastq format")
+            return num_lines // 4
+        raise InvalidInputFileError("Unable to recognize file format")
+
+
+reads = count_reads
 
 
 def files_have_min_reads(input_files, min_reads):
     """ Checks whether fa/fasta/fq/fastq have the minimum number of reads.
     Pipeline steps can use this method for input validation.
     """
     for input_file in input_files:
-        num_reads = _count_reads_via_wc(input_file, max_reads=min_reads)
+        num_reads = count_reads(input_file)
```

Comment:
Nice rename, strange to have an implementation detail like that in a function name.